### DISCUSSION TITLE
Flask displaying old / cached version of the app

### DISCUSSION THREAD

**Discussion by s2t2** [pallets/flask#4875] [Q&A]
A handful of my students are encountering an issue whereby they are developing a flask app locally, and when they make some new changes to the code and restart the local web server, it doesn't show the updated changes, but rather what appears to be an older state of the web app code.

I have confirmed they are saving the files properly and restarting the web app, and their code looks identical to mine and other students whose apps are working locally. The weird thing is the same code and usage command works for me and many of my students, but not for a few of them.

To replicate the bug, we are using the following application factory function in the "web_app/\_\_init\_\_.py" file, and running the app with `FLASK_APP=web_app flask run` command:

```py
# this is the "web_app/__init__.py" file...

from flask import Flask

from web_app.routes.home_routes import home_routes

def create_app():
    app = Flask(__name__)
    app.register_blueprint(home_routes)
    return app

if __name__ == "__main__":
    my_app = create_app()
    my_app.run(debug=True)
```

We tried using the configuration variable `app.config["TEMPLATES_AUTO_RELOAD"] = True` in the application factory function, but it has no effect.

We also tried removing / commenting out the code in the main conditional at the bottom of the "web_app/\_\_init\_\_.py" file, upgrading Flask version to 2.2.1, and using the new flask run command `flask --app web_app --debug run`. But that shows the same behavior as before.

Please advise on what structure and configuration variables we should be using to fix this issue. Thanks!

Environment:

- Python version: 3.8
- Flask version: various versions, including a 1.X version as well as 2.2.1
- I believe most if not all students who are experiencing this issue are on Windows. Some are using the Edge browser, but some are using Chrome. We tried hard refreshes of the page, but that didn't work either.

Update:

I don't think this issue is related to a previous version of the web server still running, because we don't see any issues related to the address being in use.

**Comment by davidism** [comment]
I vaguely recall that on Windows it's sometimes possible to run the server without stopping a previous instance, and you won't get an "address in use" error you'll just keep getting the old instance. Check that only one instance of the dev server is running.

**Reply by s2t2** [reply]
Thanks @davidism for the quick reply! 

This sounds like it could be the issue (I would normally be looking out for that "address in use" error but if it is not getting produced that could definitely be the issue in question). 

What command(s) would you recommend on Windows to check which / how many instances are running? Are those these:

```sh
# windows:
netstat -ano | findstr 5000
```
https://flask.palletsprojects.com/en/2.2.x/server/#address-already-in-use

**Reply by s2t2** [reply]
We can run the following command from powershell, providing the number from the netstat command:

```sh
netstat -ano | findstr :5000

# where 82932 is and example process number found from the previous command
stop-process 82932
```


https://stackoverflow.com/a/69047484/670433



This begs the question though, why are all these additional processes being spawned and not cleaned up. Is that something we could / should fix in Flask?
